tests/bsim/bluetooth/audio/src/csip_set_coordinator_test.c

```diff
@@ -244,6 +253,318 @@ static void ordered_access(const struct bt_csip_set_coordinator_set_member **mem
+					BT_CONN_LE_CREATE_CONN,
+					BT_LE_CONN_PARAM_DEFAULT,
+					&conns[i]);
+		if (err != 0) {
+			FAIL("Failed to connect to %s: %d\n", addr, err);
+			return;
+		}
+
+		printk("Connected to %s\n", addr);
+		WAIT_FOR_FLAG(flag_connected);
+		connected_member_count++;
+
+		discovered = false;
+		printk("Doing discovery on member[%u]", i);
+		err = bt_csip_set_coordinator_discover(conns[i]);
+		if (err != 0) {
+			FAIL("Failed to initialize set coordinator for connection %d\n",
+			      err);
+			return;
+		}
+
+		WAIT_FOR_COND(discovered);
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(locked_members); i++) {
+		locked_members[i] = set_members[i];
+	}
+
+	if (inst->info.rank != 0U) {
+		ordered_access(locked_members, connected_member_count, false);
+	}
+
+	/* Now disconnect one of the members */
+	bt_conn_disconnect(conns[0], BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+	printk("Member disconnected");
+
+	if (inst->info.lockable) {
+		printk("Locking set\n");
+		/* This should fail */
+		err = bt_csip_set_coordinator_lock(locked_members, connected_member_count,
+						   &inst->info);
+		if (err != 0) {
+			PASS("Failed to do set coordinator lock (%d)\n", err);
+			for (uint8_t i = 1; i < members_found; i++) {
+				printk("Disconnecting member[%u] (%s)", i, addr);
+				err = bt_conn_disconnect(conns[i],
+							 BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+				(void)memset(&set_members[i], 0, sizeof(set_members[i]));
+
+				return;
+			}
+		}
+	}
+	PASS("All members disconnected\n");
+}
+
+static void test_reconnect(void)
+{
+	int err;
+	char addr[BT_ADDR_LE_STR_LEN];
+	const struct bt_csip_set_coordinator_set_member *locked_members[CONFIG_BT_MAX_CONN];
+	uint8_t connected_member_count = 0;
+
+	err = bt_enable(NULL);
+	if (err != 0) {
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Audio Client: Bluetooth initialized\n");
+
+	bt_csip_set_coordinator_register_cb(&cbs);
+	k_work_init_delayable(&discover_members_timer,
+			      discover_members_timer_handler);
+	bt_le_scan_cb_register(&csip_set_coordinator_scan_callbacks);
+
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, NULL);
+	if (err != 0) {
+		FAIL("Scanning failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Scanning successfully started\n");
+
+	WAIT_FOR_COND(members_found == 1);
+
+	printk("Stopping scan\n");
+	err = bt_le_scan_stop();
+	if (err != 0) {
+		FAIL("Could not stop scan");
+		return;
+	}
+
+	bt_addr_le_to_str(&addr_found[0], addr, sizeof(addr));
+	err = bt_conn_le_create(&addr_found[0], BT_CONN_LE_CREATE_CONN,
+				BT_LE_CONN_PARAM_DEFAULT, &conns[0]);
+	if (err != 0) {
+		FAIL("Failed to connect to %s: %d\n", err);
+		return;
+	}
+	printk("Connecting to %s\n", addr);
+
+	WAIT_FOR_FLAG(flag_connected);
+	connected_member_count++;
+
+	err = bt_csip_set_coordinator_discover(conns[0]);
+	if (err != 0) {
+		FAIL("Failed to initialize set coordinator for connection %d\n",
+		     err);
+		return;
+	}
+
+	WAIT_FOR_COND(discovered);
```

Comment:
A lot of this can be moved to a common function for the tests

Reply:
A lot of the code below seems like it can also be moved to some common functions to reduce code duplication